### PR TITLE
OCPBUGS-26016: Changed the s3:HeadBucket entry to s3:ListBucket

### PR DESCRIPTION
**Pull request by dfitzmau**
Version(s):
4.14+

Issue:
[OCPBUGS-26016](https://issues.redhat.com/browse/OCPBUGS-26016)

Link to docs preview:
[Required AWS permissions for the IAM user](https://70926--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-account#installation-aws-permissions_installing-aws-account)

- [x] SME has approved this change.
- [x] QE has approved this change.

Additional resources:
* [permissions.go for AWS](https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/aws/permissions.go)
* [Original permission published doc](https://docs.openshift.com/container-platform/4.15/installing/installing_aws/installing-aws-user-infra.html#installation-aws-permissions_installing-aws-user-infra)

### CODE REVIEW

File: modules/installation-aws-permissions.adoc [hunk 1]
```diff
@@ -185,8 +189,9 @@ If you have not created a load balancer in your AWS account, the IAM user also r
 * `s3:GetBucketCors`
 * `s3:GetBucketLocation`
 * `s3:GetBucketLogging`
-* `s3:GetBucketPolicy`
 * `s3:GetBucketObjectLockConfiguration`
+* `s3:GetBucketPolicy`
+* `s3:GetReplicationConfiguration` #listed twice
```
Comment (r4f4) [inline modules/installation-aws-permissions.adoc:194]: One of the instances of this permission can be removed. No point in listing it twitce in the same section.

File: modules/installation-aws-permissions.adoc [hunk 2]
```diff
@@ -264,19 +269,19 @@ If you use an existing VPC, your account does not require these permissions to d
...
-* `iam:ListAccessKeys`
-* `iam:PutUserPolicy`
-* `iam:TagUser`
-* `s3:PutBucketPublicAccessBlock`
-* `s3:GetBucketPublicAccessBlock`
-* `s3:PutLifecycleConfiguration`
-* `s3:HeadBucket`
-* `s3:ListBucketMultipartUploads`
-* `s3:AbortMultipartUpload`
+* `iam:DeleteAccessKey` # Also listed in the Required permissions to delete base cluster resources table, which matches the go file.
```
Comment (r4f4) [inline modules/installation-aws-permissions.adoc:272]: I guess it's fine to repeat some perms in different sections because customers can use different roles for install vs destroy, for example.
Not sure why some `iam:Delete*` perms are need for manifests creation, though.

**Review by r4f4 — COMMENTED** [review]
@patrickdillon do you know who could help review the required permissions in the docs? I suspect that those not listed in the [Installer](https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/aws/permissions.go) might be needed by some cluster operators.

File: modules/installation-aws-permissions.adoc (hunk 1 above)
Comment (dfitzmau) [reply modules/installation-aws-permissions.adoc:194]: Thanks. Removed.

File: modules/installation-aws-permissions.adoc [hunk 3]
```diff
@@ -70,8 +73,8 @@ cluster, the IAM user requires the following permissions:
 [%collapsible]
 ====
 * `ec2:AllocateAddress`
-* `ec2:AssociateAddress`
 * `ec2:AssociateDhcpOptions`
+* `ec2:AssociateAddress`
```
Comment (patrickdillon) [inline modules/installation-aws-permissions.adoc:77]: Hm why this change? It looks like all it is doing is making the list non-alphabetically ordered?

**Review by patrickdillon — COMMENTED** [review]
> @patrickdillon do you know who could help review the required permissions in the docs? I suspect that those not listed in the [Installer](https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/aws/permissions.go) might be needed by some cluster operators.

@makentenza would have a good idea who might know about this.  I'm not sure how the minimal crededential requirements (e.g. for "create manifests") were determined. I suspect by QE, but mak may have a better idea.

File: modules/installation-aws-permissions.adoc (hunk 3 above)
Comment (dfitzmau) [reply modules/installation-aws-permissions.adoc:77]: Hi @patrickdillon . Correct. I just changed the order to match https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/aws/permissions.go . Mkes comparing the permissions.go fiel and .adoc list a bit quicker.

**Comment by dfitzmau** [comment]
The .adoc file does not list the following permissions that reside in the permissions.go file: 
Screenshot from 2024-02-12 12-12-43

**Review by faermanj — APPROVED** [review]
Looks good, thanks for the update

**Review by bburt-rh — APPROVED** [review]
LGTM. 

Also, does this change also apply to the 4.16 branch? If the PR is against `main`, it should also be applicable to 4.16. Otherwise, this should be 2 PRs, one for 4.14 and one for 4.15.

**Review by yunjiang29 — COMMENTED** [review]
@dfitzmau I have some comments below.

And for `s3:HeadBucket`, it looks like it is still required in 4.16:
```
...
 INFO[2024-04-02T03:26:10Z] Resolved release latest to registry.ci.openshift.org/ocp/release:4.16.0-0.nightly-2024-04-01-213440 
...
level=warning msg=Action not allowed with tested creds action=iam:DeleteUserPolicy
level=warning msg=Tested creds not able to perform all requested actions
level=warning msg=Action not allowed with tested creds action=s3:HeadBucket
level=warning msg=Tested creds not able to perform all requested actions
level=fatal msg=failed to fetch Cluster: failed to fetch dependency of "Cluster": failed to generate asset "Platform Permissions Check": validate AWS credentials: AWS credentials cannot be used to either create new creds or use as-is
Installer exit with code 1
```
@patrickdillon any thoughts on this?

**Comment by yunjiang29** [comment]
> > @patrickdillon do you know who could help review the required permissions in the docs? I suspect that those not listed in the [Installer](https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/aws/permissions.go) might be needed by some cluster operators.
> 
> @makentenza would have a good idea who might know about this. I'm not sure how the minimal crededential requirements (e.g. for "create manifests") were determined. I suspect by QE, but mak may have a better idea.

+1 with Patrick, depending on the cluster configurations, different permissions may be required, QE can determine the exact permissions once the cluster configuration is determined

**Comment by r4f4** [comment]
> @dfitzmau I have some comments below.
> 
> And for `s3:HeadBucket`, it looks like it is still required in 4.16:
> 
> ```
> ...
>  INFO[2024-04-02T03:26:10Z] Resolved release latest to registry.ci.openshift.org/ocp/release:4.16.0-0.nightly-2024-04-01-213440 
> ...
> level=warning msg=Action not allowed with tested creds action=iam:DeleteUserPolicy
> level=warning msg=Tested creds not able to perform all requested actions
> level=warning msg=Action not allowed with tested creds action=s3:HeadBucket
> level=warning msg=Tested creds not able to perform all requested actions
> level=fatal msg=failed to fetch Cluster: failed to fetch dependency of "Cluster": failed to generate asset "Platform Permissions Check": validate AWS credentials: AWS credentials cannot be used to either create new creds or use as-is
> Installer exit with code 1
> ```
> 
> @patrickdillon any thoughts on this?

@yunjiang29 in the docs, aws says the permission is `s3:ListBucket` https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadBucket.html.
The error you're seeing is coming from the CCO which needs to be fixed: https://github.com/openshift/cloud-credential-operator/blob/master/pkg/aws/utils.go#L57

**Comment by r4f4** [comment]
https://github.com/openshift/cloud-credential-operator/pull/690

File: modules/installation-aws-permissions.adoc [hunk 4]
```diff
@@ -289,6 +310,7 @@ If you are managing your cloud provider credentials with mint mode, the IAM user
 [%collapsible]
 ====
 * `ec2:DescribeInstanceTypeOfferings`
+* `ec2:DescribeInstanceType`
```
[the remark this answers is not in the document]
Comment (dfitzmau) [reply modules/installation-aws-permissions.adoc:313]: Removed but are the following valid:

* `ec2:DescribeInstanceTypeOfferings`
* `ec2:DescribeInstanceTypes`

File: modules/installation-aws-permissions.adoc [hunk 5]
```diff
@@ -256,6 +265,19 @@ If you use an existing VPC, your account does not require these permissions to d
 =====
 ====
 
+.Required Key Management Service (KMS) permissions
```
[the remark this answers is not in the document]
Comment (dfitzmau) [reply modules/installation-aws-permissions.adoc:268]: Thanks, @yunjiang29 . How about:

> Permissions only required if you defined a custom Key Management Service (KMS) key

File: modules/installation-aws-permissions.adoc [hunk 6]
```diff
@@ -99,16 +99,25 @@ If you use an existing VPC, your account does not require these permissions for
 * `elasticloadbalancing:ApplySecurityGroupsToLoadBalancer`
 * `elasticloadbalancing:AttachLoadBalancerToSubnets`
 * `elasticloadbalancing:ConfigureHealthCheck`
+* `elasticloadbalancing:CreateListener`
```
[the remark this answers is not in the document]
Comment (dfitzmau) [reply modules/installation-aws-permissions.adoc:102]: Yeah, these were originally spearated based on the ELB versioning. The ELB v1 has more than  ELB v2 so @patrickdillon input would be valuable here for sure.

File: modules/installation-aws-permissions.adoc [hunk 7]
```diff
@@ -266,18 +288,17 @@ If you use an existing VPC, your account does not require these permissions to d
 [%collapsible]
 ====
 * `iam:DeleteAccessKey`
-* `iam:DeleteUser`
 * `iam:DeleteUserPolicy`
```
[the remark this answers is not in the document]
Comment (dfitzmau) [reply modules/installation-aws-permissions.adoc]: Updated.

File: modules/installation-aws-permissions.adoc [hunk 8]
```diff
@@ -204,7 +214,6 @@ If you have not created a load balancer in your AWS account, the IAM user also r
 .S3 permissions that cluster Operators require
 [%collapsible]
 ====
-* `s3:DeleteObject`
```
[the remark this answers is not in the document]
Comment (r4f4) [reply modules/installation-aws-permissions.adoc]: @yunjiang29 with aws/terraform, isn't the `s3:DeleteObject` perm only needed for cluster destroy? I don't think it's needed for cluster install.
However, that's not true for CAPI/CAPA, which can delete s3 objects during machine reconcile and at the end of provisioning.

File: modules/installation-aws-permissions.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (patrickdillon) [reply modules/installation-aws-permissions.adoc:102]: This is the original PR (and it links to the BZ) that called for separating ELB & ELBv2. From what I can see in the bug, it does not make sense to split up the ELB permissions into two groups. Instead, imho, it would be better to include all ELB permissions under the same heading, and make sure that it includes all relevant permissions so no warnings are generated for either ELB or ELBv2 resources.

**Comment by patrickdillon** [comment]
> @dfitzmau I have some comments below.
> 
> And for `s3:HeadBucket`, it looks like it is still required in 4.16:
> 
> ```
> ...
>  INFO[2024-04-02T03:26:10Z] Resolved release latest to registry.ci.openshift.org/ocp/release:4.16.0-0.nightly-2024-04-01-213440 
> ...
> level=warning msg=Action not allowed with tested creds action=iam:DeleteUserPolicy
> level=warning msg=Tested creds not able to perform all requested actions
> level=warning msg=Action not allowed with tested creds action=s3:HeadBucket
> level=warning msg=Tested creds not able to perform all requested actions
> level=fatal msg=failed to fetch Cluster: failed to fetch dependency of "Cluster": failed to generate asset "Platform Permissions Check": validate AWS credentials: AWS credentials cannot be used to either create new creds or use as-is
> Installer exit with code 1
> ```
> 
> @patrickdillon any thoughts on this?

HeadBucket is an action--not a required permission--which the installer is testing to see if the creds have sufficient permissions to perform. According [to this](https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadBucket.html)  `s3:ListBucket` is the required permission for HeadBucket. `s3:ListBucket` is in the standard list of permissions, so I would assume you have it? @yunjiang29 can you confirm these creds do have `s3:ListBucket`? If they do, it sounds like there may be a bug in the installer or CCO code that the installer uses.

**Comment by r4f4** [comment]
> 
> HeadBucket is an action--not a required permission--which the installer is testing to see if the creds have sufficient permissions to perform. According [to this](https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadBucket.html) `s3:ListBucket` is the required permission for HeadBucket. `s3:ListBucket` is in the standard list of permissions, so I would assume you have it? @yunjiang29 can you confirm these creds do have `s3:ListBucket`? If they do, it sounds like there may be a bug in the installer or CCO code that the installer uses.

@patrickdillon There was. It was fixed in https://github.com/openshift/cloud-credential-operator/pull/690. And on https://github.com/openshift/installer/pull/8233 I'm bumping CCO in the Installer to bring in the fix.

File: modules/installation-aws-permissions.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (dfitzmau) [reply modules/installation-aws-permissions.adoc:102]: Thanks. Updated. I remove the ELBv2 table but cross-referenced the deleted permissions against the ELBv1 table to ensure no replica permissions were deleted.

File: modules/installation-aws-permissions.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (yunjiang29) [reply modules/installation-aws-permissions.adoc]: @dfitzmau  I'm not familiar with the doc convention, but we can refer to an existing section `Optional permissions ... `, like:
```
Optional permissions for the cluster owner account when installing a cluster on a shared VPC
```

File: modules/installation-aws-permissions.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (yunjiang29) [reply modules/installation-aws-permissions.adoc:102]: > Instead, imho, it would be better to include all ELB permissions under the same heading

+1

File: modules/installation-aws-permissions.adoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (yunjiang29) [reply modules/installation-aws-permissions.adoc]: > However, that's not true for CAPI/CAPA, which can delete s3 objects during machine reconcile and at the end of provisioning.

@r4f4 do you mean the bootstrap bucket can be deleted even no `s3:DeleteObject` permission attached to the installer?

[the remark this answers is not in the document]
Comment (r4f4) [reply modules/installation-aws-permissions.adoc]: No, I just meant that for CAPA we always need the s3:DeleteObject because ignition can be deleted and recreated during the install.

**Comment by yunjiang29** [comment]
> HeadBucket is an action--not a required permission--which the installer is testing to see if the creds have sufficient permissions to perform. According [to this](https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadBucket.html) `s3:ListBucket` is the required permission for HeadBucket. `s3:ListBucket` is in the standard list of permissions, so I would assume you have it? @yunjiang29 can you confirm these creds do have `s3:ListBucket`? If they do, it sounds like there may be a bug in the installer or CCO code that the installer uses.

Yes, it has `s3:ListBucket` permission.

File: modules/installation-aws-permissions.adoc [hunk 9]
```diff
@@ -266,18 +279,16 @@ If you use an existing VPC, your account does not require these permissions to d
 [%collapsible]
 ====
 * `iam:DeleteAccessKey`
```
Comment (yunjiang29) [inline modules/installation-aws-permissions.adoc:281]: `iam:DeleteAccessKey` should be removed from `.Additional IAM and S3 permissions that are required to create manifests`

File: modules/installation-aws-permissions.adoc [hunk 10]
```diff
@@ -212,19 +207,24 @@ If you have not created a load balancer in your AWS account, the IAM user also r
...
 [%collapsible]
 ====
 * `autoscaling:DescribeAutoScalingGroups`
-* `ec2:DeletePlacementGroup`
 * `ec2:DeleteNetworkInterface`
+* `ec2:DeletePlacementGroup`
 * `ec2:DeleteVolume`
 * `elasticloadbalancing:DeleteTargetGroup`
 * `elasticloadbalancing:DescribeTargetGroups`
-* `iam:DeleteAccessKey`
```
Comment (yunjiang29) [inline modules/installation-aws-permissions.adoc:226]: `iam:DeleteAccessKey` should NOT be removed from `.Required permissions to delete base cluster resources`

**Review by yunjiang29 — COMMENTED** [review]
@dfitzmau `iam:DeleteAccessKey` needs to be updated:

File: modules/installation-aws-permissions.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (yunjiang29) [reply modules/installation-aws-permissions.adoc:102]: @patrickdillon I re-visited the original [Bug 1902748](https://bugzilla.redhat.com/show_bug.cgi?id=1902748), it‘s more of an AWS issue (or work as design?).

Some permissions are V1 or V2 exclusive, from user's perspective, if the policy contains a V2-only permission, e.g. `DescribeListeners`, a warning always   appears on the page:

```
V1+V2: AddTags
V1+V2: CreateLoadBalancer
V1+V2: DeleteLoadBalancer
V1+V2: DescribeLoadBalancerAttributes
V1+V2: DescribeLoadBalancers
V1+V2: DescribeTags
V1+V2: ModifyLoadBalancerAttributes
V1: ApplySecurityGroupsToLoadBalancer
V1: AttachLoadBalancerToSubnets
V1: ConfigureHealthCheck
V1: CreateLoadBalancerListeners
V1: DeregisterInstancesFromLoadBalancer
V1: DescribeInstanceHealth
V1: RegisterInstancesWithLoadBalancer
V1: SetLoadBalancerPoliciesOfListener
V2: CreateListener
V2: CreateTargetGroup
V2: DeregisterTargets
V2: DescribeListeners
V2: DescribeTargetGroupAttributes
V2: DescribeTargetHealth
V2: ModifyTargetGroup
V2: ModifyTargetGroupAttributes
V2: RegisterTargets
```

image

File: modules/installation-aws-permissions.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (dfitzmau) [reply modules/installation-aws-permissions.adoc]: Hi @yunjiang29 . Are you suggesting that I move all the `kms` permissions to the existing table named "Optional permissions for the cluster owner account when installing a cluster on a shared VPC"?

File: modules/installation-aws-permissions.adoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (dfitzmau) [reply modules/installation-aws-permissions.adoc]: Hi @yunjiang29 and @r4f4 . Do I need to expand further on my note?

File: modules/installation-aws-permissions.adoc (hunk 10 above)
Comment (dfitzmau) [reply modules/installation-aws-permissions.adoc:226]: Readded `iam:DeleteAccessKey` permission to said section.

File: modules/installation-aws-permissions.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (dfitzmau) [reply modules/installation-aws-permissions.adoc:102]: Hi @yunjiang29 . I need to revert the changes and make sure ELB V1 and V2 are split out?

File: modules/installation-aws-permissions.adoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (r4f4) [reply modules/installation-aws-permissions.adoc]: @dfitzmau I don't think so. We can revisit it, if needed, when capi support is GA.

File: modules/installation-aws-permissions.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (yunjiang29) [reply modules/installation-aws-permissions.adoc:102]: I prefer to put them together actually, but my only concern is the user may ask the same question like [Bug 1902748](https://bugzilla.redhat.com/show_bug.cgi?id=1902748).

Two options may be applicable:
1) keep them unchanged
2) put all of them together in one section, and add a note to let user know that these permissions are not classified as ELB or ELBv2 permissions in OCP, but in AWS console

@patrickdillon do you have any better idea?

File: modules/installation-aws-permissions.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (yunjiang29) [reply modules/installation-aws-permissions.adoc]: @dfitzmau sorry, I mean re-title with something like `Optional permissions for installing cluster with custom Key Management Service (KMS) key`

[the remark this answers is not in the document]
Comment (dfitzmau) [reply modules/installation-aws-permissions.adoc]: Thanks. Updated.

File: modules/installation-aws-permissions.adoc [hunk 11]
```diff
@@ -199,6 +189,11 @@ If you have not created a load balancer in your AWS account, the IAM user also r
 * `s3:PutBucketAcl`
 * `s3:PutBucketTagging`
 * `s3:PutEncryptionConfiguration`
+
+[IMPORTANT]
+=====
+A known issue exists where the installation program requests a non-existent `s3:HeadBucket` permission. To resolve this issue, you must specify `s3:HeadBucket` for an IAM user. The `s3:ListBucket` permission remains as the required permission for all `HeadBucket` action requests. For more information, see link:https://issues.redhat.com/browse/OCPBUGS-31813[*OCPBUGS-31813*].
```
Comment (yunjiang29) [inline modules/installation-aws-permissions.adoc:195]: @dfitzmau this issue got resolved.

File: modules/installation-aws-permissions.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (dfitzmau) [reply modules/installation-aws-permissions.adoc:102]: Hi @yunjiang29 . I've updated the docs according to option 2 and Patrick's suggestion. OK to add a "LGTM"?

File: modules/installation-aws-permissions.adoc [hunk 12]
```diff
@@ -212,19 +207,25 @@ If you have not created a load balancer in your AWS account, the IAM user also r
 * `s3:PutObject`
 * `s3:PutObjectAcl`
 * `s3:PutObjectTagging`
+
+[NOTE]
+=====
+If you set the `preserveBootstrapIgnition` AWS parameter in your `install-config.yaml` file, the `s3:DeleteObject` permission is optional.
```
Comment (yunjiang29) [inline modules/installation-aws-permissions.adoc:213]: @dfitzmau  This description is correct, but currently, installer still checks if this permission exist [1], so we can add this note only if it's removed from the check list, or add a condition like [2], @patrickdillon WDYT?

[1] https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/aws/permissions.go#L177
[2] https://github.com/openshift/installer/pull/8296

Comment (yunjiang29) [reply modules/installation-aws-permissions.adoc:213]: @dfitzmau @patrickdillon  How about we drop this note, and still consider `s3:DeleteObject` as required permission, as the `preserveBootstrapIgnition` is not enabled by default.

Comment (yunjiang29) [reply modules/installation-aws-permissions.adoc:213]: @dfitzmau  I think we can hold this part until https://issues.redhat.com/browse/OCPBUGS-33509 gets resolved.

Comment (dfitzmau) [reply modules/installation-aws-permissions.adoc:213]: Removed:

> If you set the `preserveBootstrapIgnition` AWS parameter in your `install-config.yaml` file, the `s3:DeleteObject` permission is optional.

This work will continue on the following Jiras:

- https://issues.redhat.com/browse/OCPBUGS-33661
- https://issues.redhat.com/browse/OCPBUGS-33662
